_posts/en/newsletters/2019-04-02-newsletter.md

```diff
@@ -0,0 +1,269 @@
+  sees enough additional adoption in the next six months to warrant an
+  earlier switch, it will instead default to bech32 receiving addresses
+  in Bitcoin Core 0.19.  Legacy address will continue to be provided
+  if the user requests them in the GUI or by RPC, and anyone who doesn't
+  want the update will be able to configure their default address type.
+  (Similarly, pioneering users who want to change their default now may
+  set the `addresstype=bech32` configuration option in any Bitcoin Core
+  release from 0.16.0 up.)
+
+## Bech32 sending support
+
+*Week 3 of 24.  From now until the second anniversary of the segwit soft
+fork lock-in on 24 August 2019, the Optech Newsletter will contain this
+weekly section that provides information to help developers and
+organizations implement bech32 sending support---the ability to pay
+native segwit addresses.  This [doesn't require implementing
+segwit][bech32 easy] yourself, but it does allow the people you pay to
+access all of segwit's multiple benefits.*
+
+In a [previous week][bech32 easy], we discussed how small the
+differences are between creating the output for a legacy address versus
+a native segwit address.  In that section we simply pointed you towards
+the [bech32 reference libraries][] and told you that you'd get two
+values back.  In this week, we walkthrough the exact steps of using the
+Python reference library so you can see how little work this is.  We
+start by importing the library:
+
+```python3
+import segwit_addr
+```
+
+Bech32 addresses have a Human-Readable Part (HRP) that indicates what
+network the address is for.  These are the first few characters of the
+address and are separated from the data part of the address by the
+delimiter `1`.  For example, Bitcoin testnet uses `tb` and an example
+testnet address is tb1q3w[...]g7a.  We'll set the Bitcoin mainnet HRP of
+`bc` in our code so that we can later ensure that the addresses we parse
+are for the network we expect.
+
+```python3
+HRP='bc'
+```
+
+Finally, we have a few addresses we want to check---one that should work
+and two that should fail.  (See [BIP173][] for a complete set of
+reference test vectors.)
+
+```python3
+good_address='bc1qd6h6vp99qwstk3z668md42q0zc44vpwkk824zh'
+typo_address='bc1qd6h6vp99qwstk3z669md42q0zc44vpwkk824zh'
+wrong_network_address='tb1q3wrc5yq9c300jxlfeg7ae76tk9gsx044ucyg7a'
+```
+
+Now we can simply attempt to decode each of these addresses
+
+```python3
+>>> segwit_addr.decode(HRP, good_address)
+(0, [110, 175, 166, 4, 165, 3, 160, 187, 68, 90, 209,
+     246, 218, 168, 15, 22, 43, 86, 5, 214])
+
+In [16]: segwit_addr.decode(HRP, typo_address)
+Out[16]: (None, None)
+
+In [17]: segwit_addr.decode(HRP, wrong_network_address)
+Out[17]: (None, None)
+```
+
+If we get back a None for the first value (the witness version), the
+address is invalid on our chosen network, so you want to throw an
+exception up the stack so that whatever process is interfacing with the
```

Comment:
While the Python reference library doesn't provide reason(s) for the decode failure, presumably this would be possible to do allowing for a more actionable error message to be shown to the user. If this is true, should we speak to that here? Or, maybe in a subsequent week for "improving the user experience"?

Reply:
Ha, you're on to my plan.  I was initially going to do both a "python backend" and "JS frontend" section this week, but the Python section seemed long enough.  My plan for next week's newsletter is to code up basically a version of [this](http://bitcoin.sipa.be/bech32/demo/demo.html) via a similar tutorial in the newsletter using the reference JS ECC library.  I don't want to explicitly commit to that in this week's newsletter because I haven't written it yet and so I don't know if I'll encounter any problems.